File: Classes/PSBookmarkAddViewController.m
//
//  PSBookmarkAddViewController.m
//  PocketSword
//
//  Created by Nic Carter on 10/01/11.
//  Copyright 2011 CrossWire Bible Society. All rights reserved.
//

#import "PSBookmarkAddViewController.h"
#import "globals.h"
#import "PSBookmarks.h"
#import "PSBookmarksNavigatorController.h"

@implementation PSBookmarkAddViewController

@synthesize bookAndChapterRef, verse;

- (id)initWithBookAndChapterRef:(NSString*)ref verse:(NSString*)v {
	self = [super initWithNibName:nil bundle:nil];
	if(self) {
		self.bookAndChapterRef = ref;
		self.verse = v;
	}
	return self;
}

- (void)loadView {
	tableViewController = [[[PSBookmarksAddTableViewController alloc] initWithBookAndChapterRef:bookAndChapterRef andVerse:verse] retain];
	containingNavigationController = [[[UINavigationController alloc] initWithRootViewController:tableViewController] retain];
	self.view = containingNavigationController.view;
}

- (void)viewDidUnload {
	[super viewDidUnload];
	[tableViewController release];
	tableViewController = nil;
	[containingNavigationController release];
	containingNavigationController = nil;
}

- (void)dealloc {
	[super dealloc];
	[tableViewController release];
	[containingNavigationController release];
	self.bookAndChapterRef = nil;
	self.verse = nil;
}


@end

@implementation PSBookmarksAddTableViewController

@synthesize bookAndChapterRef, verse, folder;

#pragma mark -
#pragma mark Initialization

- (id)initWithBookAndChapterRef:(NSString*)ref andVerse:(NSString*)v {
	self = [super initWithStyle:UITableViewStyleGrouped];
	if(self) {
		self.bookAndChapterRef = ref;
		self.verse = v;
		self.folder = nil;
	}
	return self;
}

/*
- (id)initWithStyle:(UITableViewStyle)style {
    // Override initWithStyle: if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization.
    }
    return self;
}
*/


#pragma mark -
#pragma mark View lifecycle


- (void)viewDidLoad {
    [super viewDidLoad];

	descriptionTextField = [[UITextField alloc] initWithFrame:CGRectMake(20,12,260,25)];
	[descriptionTextField setPlaceholder:@""];
	descriptionTextField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
	//descriptionTextField.delegate = self;
	descriptionTextField.keyboardType = UIKeyboardTypeDefault;
	descriptionTextField.returnKeyType = UIReturnKeyDone;
	
	self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSave target:self action:@selector(saveButtonPressed)];
	self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancelButtonPressed)];
	self.navigationItem.title = NSLocalizedString(@"VerseContextualMenuAddBookmark", @"Add Bookmark");	
	
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(folderUpdated:) name:NotificationAddBookmarkInFolder object:nil];
}

- (void)cancelButtonPressed {
	[self dismissModalViewControllerAnimated:YES];
}

- (void)saveButtonPressed {
	NSString *ref = [NSString stringWithFormat:@"%@:%@", bookAndChapterRef, verse];
	NSString *description = ref;
	if(![descriptionTextField.text isEqualToString:@""]) {
		description = descriptionTextField.text;
	}
	[PSBookmarks addBookmarkWithRef:ref name:description folderString:folder];
	[self dismissModalViewControllerAnimated:YES];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
	[self.tableView reloadData];
}

/*
- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
}
*/
/*
- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
}
*/
/*
- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
}
*/
/*
// Override to allow orientations other than the default portrait orientation.
- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Return YES for supported orientations.
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}
*/

- (void)folderUpdated:(NSNotification *)notification
{
	DLog(@"%@", [notification object]);
	self.folder = [notification object];
	[self.tableView reloadSections:[NSIndexSet indexSetWithIndex:2] withRowAnimation:UITableViewRowAnimationFade];
	[self.navigationController popToViewController:self animated:YES];
}


#pragma mark -
#pragma mark Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    return 3;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
    return 1;
}


- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
	switch (section) {
		case 0:
			return @"Verse";
		case 1:
			return @"Description";
		case 2:
			return @"Folder";
		default:
			break;
	}
	return @"";
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSString *CellIdentifier = [NSString stringWithFormat:@"Cell-%d", indexPath.section];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier] autorelease];//UITableViewCellStyleValue1
		if(indexPath.section == 1) {
			[cell addSubview:descriptionTextField];
		}
    }
    
    // Configure the cell...
	switch (indexPath.section) {
		case 0:
			cell.textLabel.text = [NSString stringWithFormat:@"%@:%@", bookAndChapterRef, verse];
			cell.selectionStyle = UITableViewCellSelectionStyleNone;
			break;
		case 1:
			[descriptionTextField setPlaceholder:[NSString stringWithFormat:@"%@:%@", bookAndChapterRef, verse]];
			cell.selectionStyle = UITableViewCellSelectionStyleNone;
			break;
		case 2:
			if(self.folder) {
				cell.textLabel.lineBreakMode = UILineBreakModeHeadTruncation;
				cell.textLabel.text = [folder stringByReplacingOccurrencesOfString:PSFolderSeparatorString withString:@"/"];
			} else {
				cell.textLabel.text = NSLocalizedString(@"BookmarksTitle", @"");
			}
			cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
			break;
		default:
			break;
	}
    
    return cell;
}


/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/


/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source.
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }   
    else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view.
    }   
}
*/


/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/


/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/


#pragma mark -
#pragma mark Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	if(indexPath.section == 1) {
		[descriptionTextField becomeFirstResponder];
	} else if(indexPath.section == 2) {
		PSBookmarksNavigatorController *bnc = [[PSBookmarksNavigatorController alloc] initWithBookmarkFolder:[PSBookmarks getBookmarkFolderForFolderString:folder] parentFolders:folder isAddingBookmark:YES];
		[self.navigationController pushViewController:bnc animated:YES];
		[bnc release];
	}
}


#pragma mark -
#pragma mark Memory management

- (void)didReceiveMemoryWarning {
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Relinquish ownership any cached data, images, etc. that aren't in use.
}

- (void)viewDidUnload {
    // Relinquish ownership of anything that can be recreated in viewDidLoad or on demand.
    // For example: self.myOutlet = nil;
	[[NSNotificationCenter defaultCenter] removeObserver:self forKeyPath:NotificationAddBookmarkInFolder];
	[descriptionTextField release];
	descriptionTextField = nil;
}


- (void)dealloc {
    [super dealloc];
	self.bookAndChapterRef = nil;
	self.verse = nil;
	self.folder = nil;
}


@end

File: Classes/PSBookmarkAddViewController.swift
//
//  PSBookmarkAddViewController.swift
//  PocketSword
//
//  Ported to Swift (Wave 2). Grouped form for adding or editing a bookmark:
//    add mode (3 sections):  verse / description / folder
//    edit mode (5 sections): ref / description / folder / created / last-accessed
//  The right nav-bar item is a Save button; in add mode a Cancel button is also
//  shown on the left.
//
//  Behaviour preserved from PSBookmarkAddViewController.{h,mm}. Consumes the Wave 1
//  Swift PSBookmark / PSBookmarkFolder / PSBookmarks model and the Swift
//  PSBookmarksNavigatorController (created when the user taps the folder row).
//  The remaining Obj-C++ caller (PSModuleViewController.mm) creates this VC via
//  -initWithBookmarkToEdit:parentFolders: and
//  -initWithBookAndChapterRef:andVerse: through PocketSword-Swift.h, so the public
//  initializers keep their original selectors and the class keeps its Obj-C name
//  PSBookmarksAddTableViewController.
//
//  Originally created by Nic Carter on 10/01/11.
//  Copyright 2011 CrossWire Bible Society. All rights reserved.
//

import UIKit

@objc(PSBookmarksAddTableViewController)
final class PSBookmarksAddTableViewController: UITableViewController, UITextFieldDelegate {

    @objc var bookAndChapterRef: String?
    @objc var verse: String?
    @objc var folder: String?
    @objc var originalFolder: String?
    @objc var bookmarkBeingEdited: PSBookmark?

    private var descriptionTextField: UITextField!

    // MARK: - Initialization

    // used to add a bookmark:
    @objc(initWithBookAndChapterRef:andVerse:)
    init(bookAndChapterRef ref: String?, andVerse v: String?) {
        super.init(style: .grouped)
        self.bookAndChapterRef = ref
        self.verse = v
        self.folder = nil
        self.originalFolder = nil
        self.bookmarkBeingEdited = nil
    }

    // used to edit a bookmark:
    @objc(initWithBookmarkToEdit:parentFolders:)
    init(bookmarkToEdit: PSBookmark?, parentFolders folders: String?) {
        super.init(style: .grouped)
        self.bookAndChapterRef = nil
        self.verse = nil
        self.folder = folders
        self.originalFolder = folders
        self.bookmarkBeingEdited = bookmarkToEdit
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - View lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        var fieldFrames = CGRect(x: 20, y: 12, width: 280, height: 25)
        if PSResizing.iPad() {
            // different frames for the iPad
            fieldFrames = CGRect(x: 60, y: 12, width: 560, height: 25)
        }

        descriptionTextField = UITextField(frame: fieldFrames)
        descriptionTextField.placeholder = ""
        descriptionTextField.autocapitalizationType = .sentences
        descriptionTextField.delegate = self
        descriptionTextField.keyboardType = .default
        descriptionTextField.returnKeyType = .done
        if let bookmarkBeingEdited = bookmarkBeingEdited {
            descriptionTextField.text = bookmarkBeingEdited.name
        }

        let saveButton = UIBarButtonItem(barButtonSystemItem: .save, target: self, action: #selector(saveButtonPressed))
        navigationItem.rightBarButtonItem = saveButton
        if bookmarkBeingEdited == nil {
            let cancelButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelButtonPressed))
            navigationItem.leftBarButtonItem = cancelButton
            navigationItem.title = NSLocalizedString("VerseContextualMenuAddBookmark", comment: "Add Bookmark")
        } else {
            navigationItem.title = NSLocalizedString("BookmarkEditBookmarkTitle", comment: "Edit Bookmark")
        }

        NotificationCenter.default.addObserver(self, selector: #selector(folderUpdated(_:)), name: NSNotification.Name(NotificationAddBookmarkInFolder), object: nil)
    }

    @objc func cancelButtonPressed() {
        // this button only exists if we're adding a bookmark, so it's ok to only do this.
        dismiss(animated: true, completion: nil)
    }

    @objc func saveButtonPressed() {
        var valid = true
        if let bookmarkBeingEdited = bookmarkBeingEdited, bookmarkBeingEdited.name == descriptionTextField.text {
            // tis ok.
        } else {
            for case let childFolder as PSBookmarkFolder in (PSBookmarks.getBookmarkFolder(forFolderString: self.folder).children ?? []) {
                if childFolder.name == descriptionTextField.text {
                    valid = false
                    break
                }
            }
        }
        if !valid {
            let alert = UIAlertController(title: NSLocalizedString("BookmarksDuplicateBookmarkTitle", comment: ""), message: NSLocalizedString("BookmarksDuplicateBookmarkMessage", comment: ""), preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: NSLocalizedString("Ok", comment: "Ok"), style: .cancel) { [weak self] _ in
                self?.descriptionTextField.becomeFirstResponder()
            })
            present(alert, animated: true, completion: nil)
            return
        }

        if let bookmarkBeingEdited = bookmarkBeingEdited {
            var description = descriptionTextField.text ?? ""
            if description == "" {
                description = bookmarkBeingEdited.ref ?? ""
            }
            let newBookmark = PSBookmark(name: description, dateAdded: bookmarkBeingEdited.dateAdded, dateLastAccessed: Date(), bibleReference: bookmarkBeingEdited.ref)
            PSBookmarks.deleteBookmark(bookmarkBeingEdited.name ?? "", fromFolderString: self.originalFolder)
            _ = PSBookmarks.addBookmarkObject(newBookmark, withFolderString: self.folder)

            let fullRef = (bookmarkBeingEdited.ref ?? "").components(separatedBy: ":")
            let ref = fullRef.first ?? ""
            if PSModuleController.createRefString(PSModuleController.getCurrentBibleRef()) == ref {
                NotificationCenter.default.post(name: NSNotification.Name(NotificationBookmarksChanged), object: nil)
            }

            navigationController?.popViewController(animated: true)
        } else {
            let ref = "\(bookAndChapterRef ?? ""):\(verse ?? "")"
            var description = ref
            if let text = descriptionTextField.text, text != "" {
                description = text
            }
            _ = PSBookmarks.addBookmark(withRef: ref, name: description, folderString: folder)
            if PSModuleController.createRefString(PSModuleController.getCurrentBibleRef()) == bookAndChapterRef {
                NotificationCenter.default.post(name: NSNotification.Name(NotificationBookmarksChanged), object: nil)
            }
            dismiss(animated: true, completion: nil)
        }
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        tableView.reloadData()
    }

    @objc func folderUpdated(_ notification: Notification) {
        //DLog(@"%@", [notification object]);
        self.folder = notification.object as? String
        tableView.reloadSections(IndexSet(integer: 2), with: .fade)
        navigationController?.popToViewController(self, animated: true)
    }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return PSResizing.supportedInterfaceOrientations()
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // Return the number of sections.
        if bookmarkBeingEdited != nil {
            return 5
        } else {
            return 3
        }
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // Return the number of rows in the section.
        return 1
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0:
            return NSLocalizedString("BookmarksAddBookmarkVerseTitle", comment: "")
        case 1:
            return NSLocalizedString("BookmarksAddBookmarkDescriptionTitle", comment: "")
        case 2:
            return NSLocalizedString("BookmarksAddBookmarkFolderTitle", comment: "")
        case 3:
            return NSLocalizedString("BookmarksCreatedTitle", comment: "")
        case 4:
            return NSLocalizedString("BookmarksLastAccessedTitle", comment: "")
        default:
            break
        }
        return ""
    }

    // Customize the appearance of table view cells.
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        let cellIdentifier = "Cell-\(indexPath.section)"

        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: cellIdentifier)
            if indexPath.section == 1 {
                cell?.addSubview(descriptionTextField)
            }
        }

        // Configure the cell...
        switch indexPath.section {
        case 0:
            if let bookmarkBeingEdited = bookmarkBeingEdited {
                cell?.textLabel?.text = bookmarkBeingEdited.ref
            } else {
                cell?.textLabel?.text = "\(bookAndChapterRef ?? ""):\(verse ?? "")"
            }
            cell?.selectionStyle = .none
        case 1:
            if let bookmarkBeingEdited = bookmarkBeingEdited {
                descriptionTextField.placeholder = bookmarkBeingEdited.ref
            } else {
                descriptionTextField.placeholder = "\(bookAndChapterRef ?? ""):\(verse ?? "")"
            }
            cell?.selectionStyle = .none
        case 2:
            if let folder = folder {
                cell?.textLabel?.lineBreakMode = .byTruncatingHead
                cell?.textLabel?.text = folder.replacingOccurrences(of: AppConstants.folderSeparatorString, with: "/")
            } else {
                cell?.textLabel?.text = NSLocalizedString("BookmarksTitle", comment: "")
            }
            cell?.accessoryType = .disclosureIndicator
        case 3:
            let dateFormatter = DateFormatter()
            dateFormatter.timeStyle = .short
            dateFormatter.dateStyle = .full
            if let dateAdded = bookmarkBeingEdited?.dateAdded {
                cell?.textLabel?.text = dateFormatter.string(from: dateAdded)
            }
            cell?.selectionStyle = .none
        case 4:
            let dateFormatter = DateFormatter()
            dateFormatter.timeStyle = .short
            dateFormatter.dateStyle = .full
            if let dateLastAccessed = bookmarkBeingEdited?.dateLastAccessed {
                cell?.textLabel?.text = dateFormatter.string(from: dateLastAccessed)
            }
            cell?.selectionStyle = .none
        default:
            break
        }

        return cell!
    }

    // MARK: - Table view delegate

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.section == 1 {
            descriptionTextField.becomeFirstResponder()
        } else if indexPath.section == 2 {
            if folder == nil {
                let bnc = PSBookmarksNavigatorController(bookmarkFolder: PSBookmarks.default(), parentFolders: nil, isAddingBookmark: true)
                navigationController?.pushViewController(bnc, animated: true)
            } else {
                let components = (folder ?? "").components(separatedBy: AppConstants.folderSeparatorString)
                // push the root of our bookmarks:
                var bnc = PSBookmarksNavigatorController(bookmarkFolder: PSBookmarks.default(), parentFolders: nil, isAddingBookmark: true)
                navigationController?.pushViewController(bnc, animated: false)
                let currentFolder = NSMutableString(string: components[0])
                var i = 0
                while i < components.count {
                    var animate = false
                    if i == (components.count - 1) {
                        animate = true
                    }
                    bnc = PSBookmarksNavigatorController(bookmarkFolder: PSBookmarks.getBookmarkFolder(forFolderString: currentFolder as String), parentFolders: currentFolder as String, isAddingBookmark: true)
                    navigationController?.pushViewController(bnc, animated: animate)
                    i += 1
                    if i < components.count {
                        currentFolder.appendFormat("%@%@", AppConstants.folderSeparatorString, components[i])
                    }
                }
                self.folder = nil // reset the folder to the root.
            }
        }
    }

    // MARK: - Memory management

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}
